### ISSUE TITLE
PriorityQueue is not reordered on .clone in scala >= 2.12

### ISSUE THREAD

**Issue by mbharanya** [scala/bug#12517]
## reproduction steps

The following code builds a `PriorityQueue` with an `Elem` case class which contains the mutable member `priority`.  After that the priority is changed on one element. The `PriorityQueue` is cloned and then all elements are dequeued.

```scala

  case class Elem(var priority: Int, i: Int)

  def MyOrdering = new Ordering[Elem] {
    def compare(a : Elem, b : Elem) = a.priority.compare(b.priority)
  }

  val pq = new scala.collection.mutable.PriorityQueue[Elem]()(MyOrdering)  ++ List(Elem(1,1), Elem(0,0), Elem(2,2))

  pq.find(x => x.priority == 0) match {
    case Some(elem: Elem) => elem.priority = 3
    case None => println("Not found")
  }

  val pq2 = pq.clone

  println(pq2)
  println(pq2.dequeue())
  println(pq2.dequeue())
  println(pq2.dequeue())
```

## problem

I expected the cloning to call the reorder by calling the compare function again. This is the case for Scala <= 2.12.

Output for Scala 2.11.12:
```
PriorityQueue(Elem(3,0), Elem(2,2), Elem(1,1))
Elem(3,0)
Elem(2,2)
Elem(1,1)
```

Output for Scala 3.1.0: 
```
PriorityQueue(Elem(2,2), Elem(3,0), Elem(1,1))
Elem(2,2)
Elem(3,0)
Elem(1,1)
```

I would have expected the same output for both scala versions.
It is possible that this is actually a bugfix introduced in scala  >= 2.12, if so please point me in the right direction.

**Comment by Jasper-M** [comment]
This seems to have changed to improve performance: https://github.com/scala/scala/pull/5181
It also seems more like an implementation detail that you're depending on. You just shouldn't change the priority of a thing that's inside of a PriorityQueue. It's like changing the hashcode of something in a HashMap.

That said, the doc does mention some kind of reordering behavior of `clone`:

>Therefore, printing a PriorityQueue will not reveal the priority order of the elements, though the highest-priority element will be printed first. To print the elements in order, one must duplicate the PriorityQueue (by using clone, for instance) and then dequeue them

Though that sentence doesn't make much sense to me. To print in order you have to clone it and then dequeue. I would hope that dequeuing always happens in order.

**Comment by SethTisue** [comment]
@scala/collections

**Comment by mbharanya** [comment]
> This seems to have changed to improve performance: [scala/scala#5181](https://github.com/scala/scala/pull/5181) It also seems more like an implementation detail that you're depending on. You just shouldn't change the priority of a thing that's inside of a PriorityQueue. It's like changing the hashcode of something in a HashMap.
> 
> That said, the doc does mention some kind of reordering behavior of `clone`:
> 
> > Therefore, printing a PriorityQueue will not reveal the priority order of the elements, though the highest-priority element will be printed first. To print the elements in order, one must duplicate the PriorityQueue (by using clone, for instance) and then dequeue them
> 
> Though that sentence doesn't make much sense to me. To print in order you have to clone it and then dequeue. I would hope that dequeuing always happens in order. Ok no, I got it now. The cloning is not about reordering, it's just because if you only want to print the elements in order you probably don't want to actually empty your mutable PriorityQueue.

I agree that this is a strange use case. But as far as I understood it, clone copies all elements of the list, which are the new items with the changed value. So in essence (correct me if I am wrong) it should be that a new PriorityQueue is built with the same elements and therefore should also return the items with the correct ordering.

**Comment by NthPortal** [comment]
I don't really think it's fair to expect `clone` to rebuild the `PriorityQueue`. personally, I would expect `clone` to produce a copy that is exactly equivalent, which would have it be in the exact same order. also, modifying an element that's already in a `PriorityQueue` breaks all guarantees of the `PriorityQueue` behaving in a consistent and non-arbitrary way

**Comment by mbharanya** [comment]
> I don't really think it's fair to expect `clone` to rebuild the `PriorityQueue`. personally, I would expect `clone` to produce a copy that is exactly equivalent, which would have it be in the exact same order. also, modifying an element that's already in a `PriorityQueue` breaks all guarantees of the `PriorityQueue` behaving in a consistent and non-arbitrary way

I agree that `clone` should create an exact copy, but I think it is also fair to assume that a PriorityQueue will always return the right order of items according to the `Ordering` defined. I think this might be the key point here, that there are two competing assumptions.

**Comment by sjrd** [comment]
> also, modifying an element that's already in a `PriorityQueue` breaks all guarantees of the `PriorityQueue` behaving in a consistent and non-arbitrary way

This is the most important thing here. As soon as you mutate an element inside a `PriorityQueue` in a way that changes its order, that `PriorityQueue` *is corrupted*. No operation on it is specified anymore. Be it cloning it, listing its elements, removing elements, or even printing it. Doing anything with that `PriorityQueue` is unspecified.

**Comment by som-snytt** [comment]
I'll submit a doc tweak for Jasper-M's observation. It's also strange that the example is separated from the text, as rendered. There's also an Ichoran request to document the off-by-one padding in the heapified array.

I'll look for a place to add words about mutability of keys and sorted elements, but possibly it would have to be a doc macro expanded everywhere, in order to be useful or visible.

**Comment by som-snytt** [comment]
The PR doesn't yet have tweaking words about mutable keys and status of first element in the implementing array. I saw that Sedgewick has words to say on both issues, so either we defer to Sedgewick (by assuming it) or replicate it. I'm inclined to replicate, so as not to assume too much.

There is a new test for a previous PR?

I also wanted to venture into additional tweaks, but not without a benchmark, which I started elsewhere.

I don't know if a "tweaking word" is a "twerk"?
